File: Tareas/Tarea3/Parte2/Pregunta1/Parte2_Articulo2.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.fftpack import dct, idct
from PIL import Image

# Función para aplicar DCT a bloques de 8x8
def apply_dct(image):
    return dct(dct(image.T, norm='ortho').T, norm='ortho')

# Función para aplicar DCT inversa a bloques de 8x8
def apply_idct(image):
    return idct(idct(image.T, norm='ortho').T, norm='ortho')

# Función para aplicar SVD
def apply_svd(matrix):
    return np.linalg.svd(matrix, full_matrices=False)

# Función para reconstruir una matriz a partir de U, S, V
def reconstruct_matrix(u, s, vh):
    return np.dot(u, np.dot(np.diag(s), vh))

# Cargar la imagen original y la marca de agua usando PIL y convertir a escala de grises
original_image = Image.open('imagen.jpg').convert('L')
watermark = Image.open('marca.jpg').convert('L')

# Convertir imágenes de PIL a arrays de NumPy
original_image = np.array(original_image)
watermark = np.array(watermark)
watermark = np.resize(watermark, (32, 32))  # Asegúrate de que el tamaño sea el correcto

# Incrustación de la marca de agua
block_size = 8
alpha = 0.1  # Factor de ganancia

# Dividir en bloques y aplicar DCT
dct_blocks = np.array([
    apply_dct(original_image[i:i+block_size, j:j+block_size])
    for i in range(0, original_image.shape[0], block_size)
    for j in range(0, original_image.shape[1], block_size)
]).reshape(original_image.shape[0] // block_size, original_image.shape[1] // block_size, block_size, block_size)

# Extraer la matriz DC y aplicar SVD
dc_matrix = dct_blocks[:, :, 0, 0]
u, s, vh = apply_svd(dc_matrix)

# Modificar la matriz S con la marca de agua
s_watermarked = s + alpha * watermark.flatten()

# Reconstruir la matriz A y reemplazar en los bloques DCT
watermarked_dc_matrix = reconstruct_matrix(u, s_watermarked, vh)
dct_blocks[:, :, 0, 0] = watermarked_dc_matrix

# Aplicar DCT inversa para obtener la imagen marcada
watermarked_image = np.array([
    apply_idct(dct_blocks[i, j])
    for i in range(dct_blocks.shape[0])
    for j in range(dct_blocks.shape[1])
]).reshape(original_image.shape)

# Extracción de la marca de agua
extracted_watermark = (s_watermarked - s) / alpha

# Visualización de imágenes
fig, ax = plt.subplots(1, 3, figsize=(10, 5))
ax[0].imshow(original_image, cmap='gray')
ax[0].set_title('Imagen Original')
ax[1].imshow(watermarked_image, cmap='gray')
ax[1].set_title('Imagen con Marca de Agua')
ax[2].imshow(extracted_watermark.reshape(32, 32), cmap='gray')
ax[2].set_title('Marca de Agua Extraída')
plt.show()

File: Tareas/Tarea3/Parte2/Pregunta1/Parte2_Articulo2.m
pkg load image;

% Leer la imagen original y convertir a escala de grises
I = double(imread('imagen.jpg'));

% Leer y preparar la imagen de marca de agua
watermark = double(imread('marca.jpg'));

% Parámetro de ganancia
alpha = 0.05;

function [U, S, V, watermarked_image] = apply_water_mark(Image, Watermark, alpha)
    % Dividir en bloques de 8x8 y aplicar DCT
    DCT_blocks = blockproc(Image, [8 8], @(block) dct2(block));

    % Extraer valores DC y construir la matriz A
    A = blockproc(DCT_blocks, [8 8], @(block) block(1,1));

    % Aplicar SVD
    [U, S, V] = svd(A);

    % Modificar solo los valores singulares (diagonal de S)
    S_mod = S;
    S_mod(1:min(size(S), size(Watermark)), 1:min(size(S), size(Watermark))) = S_mod(1:min(size(S), size(Watermark)), 1:min(size(S), size(Watermark))) + alpha * Watermark;

    % Recomponer A modificado
    A_mod = U * S_mod * V';

    % Reemplazar los valores DC en los bloques DCT
    blocks = mat2cell(DCT_blocks, repmat(8, 1, size(Image,1)/8), repmat(8, 1, size(Image,2)/8));

    for i = 1:size(blocks,1)
        for j = 1:size(blocks,2)
            blocks{i,j}(1,1) = A_mod(i,j);
        end
    end

    DCT_blocks_mod = cell2mat(blocks);

    % Aplicar DCT inversa y obtener la imagen final
    watermarked_image = blockproc(DCT_blocks_mod, [8 8], @(block) idct2(block));
end


function watermark_extracted = remove_water_mark(Image, watermarked_image, alpha)
    % Inicializar la matriz de DCT extraída
    DCT_blocks_extracted = zeros(size(watermarked_image)) / alpha;

    block_size = 8;

    % Inicializar la marca de agua extraída
    watermark_extracted = zeros(size(Image)/block_size);

    % Procesar cada bloque de 8x8

    for i = 1:block_size:size(watermarked_image,1)
        for j = 1:block_size:size(watermarked_image,2)
            block = watermarked_image(i:i+block_size-1, j:j+block_size-1);
            DCT_block = dct2(block);

            % Recuperar la marca de agua del coeficiente DC
            watermark_extracted(ceil(i/block_size), ceil(j/block_size)) = (DCT_block(1,1) - dct2(Image(i:i+block_size-1, j:j+block_size-1))(1,1)) / alpha;
        end
    end
end


[U, S, V, watermarked_image] = apply_water_mark(I, watermark, alpha);
watermark_extracted = remove_water_mark(I, watermarked_image, alpha);



% Mostrar la imagen con marca de agua
figure;
subplot(2,2,1)
imshow(uint8(I))
title('Imagen original')

subplot(2,2,2)
imshow(uint8(watermark))
title('Marca de agua')

subplot(2,2,3)
imshow(uint8(watermarked_image))
title('Imagen con marca de agua')

subplot(2,2,4)
imshow(uint8(watermark_extracted));
title('Marca de Agua Extraída');

